fallback.go

```diff
@@ -43,22 +43,27 @@ func New(trace plugin.Handler) (f *Fallback) {
 func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
 	nw := nonwriter.New(w)
 	rcode, err := plugin.NextOrFailure(f.Name(), f.Next, ctx, nw, r)
-	if err != nil {
-		return rcode, err
-	}
-	// PROXY ALWAYS RETURN 0
-	if rcode == 0 {
-		rcode = nw.Msg.Rcode
+
+	//By default the rules_index is equal rcode, so in such way we handle the case
+	//when rcode is SERVFAIL and nw.Msg is nil, otherwise we use nw.Msg.Rcode
+	//because, for example, for the following cases like NXDOMAIN, REFUSED the rcode is 0 (returned by proxy)
+	//A proxy doesn't return 0 only in case SERVFAIL
+	rules_index := rcode
+	if nw.Msg != nil {
+		rules_index = nw.Msg.Rcode
 	}
-	if u, ok := f.rules[rcode]; ok {
+
+	if u, ok := f.rules[rules_index]; ok {
 		p := f.proxy.New(f.trace, u)
 		if p == nil {
-			return rcode, fmt.Errorf("cannot create fallback proxy")
+			return rules_index, fmt.Errorf("cannot create fallback proxy")
 		}
 		return p.ServeDNS(ctx, w, r)
 	}
-	w.WriteMsg(nw.Msg)
-	return rcode, nil
+	if nw.Msg != nil {
```

Comment:
test nw.Msg != nil is for safe.
but the logic test would be err != nil.

=> if you return an Error, then you return (SERVAIL, err) and the MSG is not written .. - I am not sure here.

Reply:
it seems it's not for safe: when we get SERVFAIL the nw.Msg is nil and if we don't have  appropriate rules, we mustn't do w.WriteMsg(nw.Msg) because of we will get SIGSEGV.

Reply:
For example, in config we have next:
fallback NXDOMAIN . 10.6.0.6 {
      policy first
      }

but we get SERVFAIL, in this case nw.Msg.

Reply:
And yea, instead of nw.Msg != nil we can check err != nil, do I need to modify it?

Reply:
I think would be good to check : err == nil && nw.Msg != nil

Reply:
ok, sure

Reply:
After reflexion, I guess you are right. if Msg != nil should be ok. (Err is only an information).